{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CRISPRCasMiner v1.5.0\n",
    "A CRISPR-Cas systems mining pipeline.\n",
    "\n",
    "`Input`: Metagenome-assembled genomes/contigs.\n",
    "\n",
    "`Output`: Known CRISPR-Cas systems & suspicious proteins adjacent to CRISPR arrays.\n",
    "\n",
    "\n",
    "*Note: before executing this notebook, please ensure that you have to save/add the following folders into your Google Drive (\"MyDrive\" folder):*\n",
    "\n",
    "- [data](https://drive.google.com/drive/folders/1WcClm_TebUuuRY_b8i6n6EuEXdR_WhT2?usp=sharing)\n",
    "- [inputs](https://drive.google.com/drive/folders/18GGlIEWYtJVTn2oBXMqbghyYQCLKelLg?usp=sharing)\n",
    "- [ccminer](https://drive.google.com/drive/folders/1uPjufxWPjEmx2yDota60qVl05dugS0OS?usp=sharing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Step.01 setup **CRISPRCasMiner** (~5m)\n",
    "%%time\n",
    "import os, time, signal\n",
    "import sys, random, string, re\n",
    "## mount google drive\n",
    "from google.colab import drive\n",
    "drive.mount('/content/drive')\n",
    "\n",
    "## install conda\n",
    "!pip install -q condacolab\n",
    "import condacolab\n",
    "condacolab.install()\n",
    "!conda update -n base -c conda-forge conda -y\n",
    "\n",
    "## install cctyper\n",
    "!conda install -c bioconda hmmer -y\n",
    "!conda install -c bioconda prodigal -y\n",
    "!conda install -c bioconda minced -y\n",
    "!conda install -c bioconda bioawk -y\n",
    "!python -m pip install 'drawSvg~=1.9'\n",
    "!python -m pip install 'cctyper~=1.8.0'\n",
    "\n",
    "## install ccminer\n",
    "!conda install -c bioconda mafft -y\n",
    "!conda install -c bioconda viennarna -y\n",
    "!conda install -c conda-forge webencodings -y\n",
    "!conda install -c bioconda seqkit -y\n",
    "!conda install -c bioconda seqtk -y\n",
    "\n",
    "## export database\n",
    "!export CCTYPER_DB='/content/drive/MyDrive/data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Step.02 run CRISPRCasMiner to generate suspicious Cas proteins (~1m)\n",
    "%%time\n",
    "\n",
    "#@markdown **Parameters** settings\n",
    "#@markdown ---\n",
    "\n",
    "from google.colab import files\n",
    "import os\n",
    "#@title Step 02.run **CRISPRCasMiner** to generate suspicious Cas proteins (~1m)\n",
    "project_name = \"test\" #@param {type:\"string\"}\n",
    "#@markdown - The `name` of your project.\n",
    "input_fna =  \"drive/MyDrive/inputs/input_test.fna\" #@param {type:\"string\"}\n",
    "if not input_fna:\n",
    "  input_fna = files.upload()\n",
    "  input_fna = list(input_fna.keys())[0]\n",
    "#@markdown - The path of metagenome-assembled genome/contig in **YOUR_GOOGLE_DRIVE_PATH**. `drive/MyDrive/inputs/input_test.fna`\n",
    "#@markdown - Or upload your own genomic data (.fasta/.fna/.fa file).\n",
    "prodigal_mode = \"meta\" #@param [\"meta\",\"single\"] {type:\"raw\"}\n",
    "#@markdown - The mode to run prodigal. `meta` or `single`.\n",
    "span = 10 #@param {type:\"raw\"}\n",
    "#@markdown - The range (`number`) of proteins around the CRISPR array that were interested, both upstream and downstream.\n",
    "sample_name, _ = os.path.splitext(os.path.basename(input_fna))\n",
    "path = project_name\n",
    "!mkdir {path}\n",
    "\n",
    "## run cctyper\n",
    "!cctyper $(pwd)/{input_fna} $(pwd)/{path}/01_cctyper --db $(pwd)/drive/MyDrive/data --prodigal {prodigal_mode} --keep_tmp\n",
    "!prodigal -i $(pwd)/{input_fna} -d $(pwd)/{path}/01_cctyper/genes.cds -p {prodigal_mode} > /dev/null\n",
    "!rm -rf $(pwd)/{path}/01_cctyper/hmmer $(pwd)/{path}/01_cctyper/spacers $(pwd)/{path}/01_cctyper/*.log $(pwd)/{path}/01_cctyper/Flank*\n",
    "\n",
    "## run ccminer\n",
    "!python $(pwd)/drive/MyDrive/ccminer/ccminer.py $(pwd)/{input_fna} $(pwd)/{path}/02_ccminer --db $(pwd)/drive/MyDrive/data --prodigal {prodigal_mode} --span {span} --keep_tmp --cctyper_path $(pwd)/{path}/01_cctyper --database_name {project_name} --name {sample_name}\n",
    "!awk '$12==\"primary\"  {printf \">%s\\n%s\\n\",$9,$13} ' $(ls */02_ccminer/out.tab) > suspicious.faa\n",
    "!rm -rf $(pwd)/{path}/02_ccminer/spacers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## show the suspicious proteins\n",
    "!cat suspicious.faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Step.03 Package and download results\n",
    "from google.colab import files\n",
    "!zip -r {path}.result.zip ./{path}/01_cctyper ./{path}/02_ccminer ./suspicious.faa\n",
    "files.download(f\"{path}.result.zip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#**Note**\n",
    "\n",
    "Two bioinformatics pipelines were executed on your metagenome-assembled data:\n",
    "\n",
    "- `cctyper`. All known CRISPR-Cas systems with certainty.\n",
    "\n",
    "- `ccminer`. All suspious proteins located near the CRISPR arrays, excluding the already known systems.\n",
    "\n",
    "\n",
    "The folder tree of output directory:\n",
    "\n",
    "```\n",
    "YOUR_PROJECT_output\n",
    "output\n",
    "├── 01_cctyper              # output of cctyper v1.8.0\n",
    "├── 02_ccminer              # output of ccminer v1.5.0\n",
    "│   ├── arguments.tab\n",
    "│   ├── crisprs_calib.tab\n",
    "│   ├── genes.tab\n",
    "│   ├── minced.out\n",
    "│   ├── nearCRISPR.txt\n",
    "│  *├── out.tab             # proteins adjacent to CRISPRs\n",
    "│   └── seqLen.tab\n",
    "└── suspicious.faa          # suspicious Cas proteins\n",
    "```\n",
    "\n",
    "If no known CRISPR-Cas systems were identified by cctyper, then pay your attention to the suspious proteins adjacent to the CRISPR array. For the output table of ccminer, `out.tab` presents the following columns:\n",
    "\n",
    "- `DB_name`. The name of YOUR_PROJECT.\n",
    "- `Sample`. Sample name of YOUR_METAGENOMITC_DATA.\n",
    "- `CRISPR`. CRISPR id.\n",
    "- `DRrepeat`.\n",
    "- `Remark`. Upstream or downstrem of the CRISPR array.\n",
    "- `Strand`.\n",
    "- `Rank`. Distance bewteen the protein and the CRISPR array.\n",
    "- `Prolen`. Protein size (aa).\n",
    "- `Protein`. Protein id.\n",
    "- `HMM`. HMMER annotation of this protein.\n",
    "- `typetag`. Types of CRISPR-Cas systems identifed by cctyper (e.g. `I-*`, `II-*`,...). `Unkown` represents no CRISPR-Cas systems with certainty identified adjacent to the CRISPR array.\n",
    "- `level`. Additional annotation of the protein:\n",
    "  - `certain`: this protein was located within or very close to a known type of CRISPR-Cas system.\n",
    "  - `putative`: this protein was located within or very close to a putative CRISPR-Cas system.\n",
    "  - `large`: the size of this protein is over 2000aa.\n",
    "  - `split`: this protein might be incomplete.\n",
    "  - `other`: this protein was not included in the above mentioned scenarios.\n",
    "  - `primary`: this protein falls under the 'other' category and has a size over 700 amino acids, indicating its potential to be a suspicious novel Cas protein.\n",
    "- `pro_seq`. Amino acid sequence of the protein.\n",
    "- `gene_seq`. CDS of the protein."
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
